File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque
def solution(board):
    dx,dy = [0,0,-1,1], [1,-1,0,0]
    answer = 0
    goal_x, goal_y = 0,0
    n,m = len(board), len(board[0])
    visited = [[0]*m for _ in range(n)]
    q = deque()
    for i in range(n):
        for j in range(m):
            if board[i][j] == 'R':
                visited[i][j] = 1
                q.append((i,j))
                break
    def bfs():
        
        while q:
            px,py = q.popleft()
            if board[px][py] == 'G':
                return visited[px][py]
            for i in range(4):
                nx,ny = px,py
                while True:
                    nx,ny = nx+dx[i],ny+dy[i]
                    if 0<=nx<n and 0<=ny<m and board[nx][ny]=='D':
                        nx -= dx[i]
                        ny -= dy[i]
                        break
                    if nx<0 or nx>=n or ny<0 or ny>=m:
                        nx -= dx[i]
                        ny -= dy[i]
                        break
                if not visited[nx][ny]:
                    visited[nx][ny] = visited[px][py] + 1
                    q.append((nx,ny))
        return -1
    answer = bfs()
    if answer >0:
        answer -=1
    return answer

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;
class Solution {
    static int n;
    static int m;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static String[] board;
    public int solution(String[] b) {
        int answer = 0;
        board = b;
        n = board.length;
        m = board[0].length();
        for(int i=0; i<n;i++){
            for(int j=0; j<m; j++){
                if(board[i].charAt(j)=='R'){
                    return bfs(i,j);
                }
            }
        }
        
        return -1;
    }
    public static int bfs(int x,int y){
        Queue<int[]> q = new LinkedList<>();
        int[][] visited =  new int[n][m];
        
        q.offer(new int[]{x,y});
        visited[x][y] = 1;
        
        
        
        while(!q.isEmpty()){
            int[] tmp = q.poll();
            
            for (int i = 0; i < 4; i++) {
            int nx = tmp[0];
            int ny = tmp[1];
            
            // 벽이나 경계를 만날 때까지 이동
            while (0 <= nx + dx[i] && nx + dx[i] < n && 0 <= ny + dy[i] && ny + dy[i] < m
                   && board[nx + dx[i]].charAt(ny + dy[i]) != 'D') {
                nx += dx[i];
                ny += dy[i];
            }
            
            if (visited[nx][ny] == 0) {
                visited[nx][ny] = visited[tmp[0]][tmp[1]] + 1;
                if (board[nx].charAt(ny) == 'G') {
                    return visited[nx][ny] - 1;
                }
                q.offer(new int[]{nx, ny});
            }
        }
    }
        return -1;
    }
}